design/design-sketch-oam.md

```diff
@@ -0,0 +1,334 @@
+          env:
+          - name: USERNAME
+            value: cooladmin
+          - name: PASSWORD
+            valueFrom:
+              secretKeyRef:
+                name: cool-frontend
+                key: password
+          - name: BACKEND_ADDRESS
+            value: https://example.org
+          livenessProbe:
+            httpGet:
+              port: 8080
+              path: /healthz
+          readinessProbe:
+            httpGet:
+              port: 8080
+              path: /healthz
+    traits:
+    # Contemporary traits model the workload kinds to which they may apply. This
+    # could be modelled as an annotation on the CRD of each Trait kind.
+    - apiVersion: core.oam.dev/v1alpha1
+      kind: ManualScalerTrait
+      spec:
+        replicaCount: 3
+  - component:
+      apiVersion: cache.crossplane.io/v1alpha1
+      kind: RedisCluster
+      spec:
+        classSelector:
+          matchLabels:
+            environment: production
+            region: europe
+        writeConnectionSecretToRef:
+          name: coolest-app-redis
+        engineVersion: "3.2"
+```
+
+Submitting the above `ApplicationController` would cause an instance of each of
+the scopes, components, and traits it templates to be created verbatim (modulo
+some generated object metadata) similar to how the `template` of a Kubernetes
+`ReplicaSet` becomes a set of `Pod` resources. For example:
+
+```yaml
+---
+apiVersion: core.oam.dev/v1alpha1
+kind: ContainerizedWorkload
+metadata:
+  # namespace and name are derived from the parent ApplicationConfiguration.
+  # An owner reference (not modelled here) represents that this scheduling scope
+  # is owned and controlled by its parent ApplicationConfiguration.
+  namespace: default
+  name: coolest-app-f92dm
+  annotations:
+    version: v1.0.0
+spec:
+  # Under this proposal the workloadType would be limited to the core OAM
+  # types, i.e. it would define the function of the resulting workload -
+  # Server, Singleton, Worker, etc. Extended workloads are a distinct kind
+  # and thus kind: ContainerizedWorkload has no workloadSettings field.
+  workloadType: core.oam.dev/v1alpha1.Server
+  osType: linux
+  containers:
+  - name: my-twitter-bot-frontend
+    image:
+      name: example/my-twitter-bot-frontend:v1.0.0
+    resources:
+      cpu:
+        required: 1.0
+      memory:
+        required: 100MB
+    ports:
+    - name: http
+      value: 8080
+    env:
+    - name: USERNAME
+      value: cooladmin
+    - name: PASSWORD
+      valueFrom:
+        secretKeyRef:
+          name: cool-frontend
+          key: password
+    - name: BACKEND_ADDRESS
+      value: https://example.org
+    livenessProbe:
+      httpGet:
+        port: 8080
+        path: /healthz
+    readinessProbe:
+      httpGet:
+        port: 8080
+        path: /healthz
+---
+apiVersion: core.oam.dev/v1alpha1
+kind: ManualScalerTrait
+metadata:
+  namespace: default
+  name: coolest-app-n3v3y
+spec:
+  replicaCount: 3
```

Comment:
> Traits should be applied into the runtime at installation/upgrade time. Traits should not be checked "lazily", but should be checked when the trait-holding components are created.

> Traits are applied in order so that should some set of traits express dependency (e.g. ingress must be set up before SSL), this can be resolved by setting an explicit ordering.

> A deployment SHOULD NOT be marked complete until all trait configurations are completed. For example, if a web server component is deployed with an autoscaler trait, the web server should not be considered "running" until (a) the web server itself is running, as determined by health checks, and (b) the autoscaler trait is running, as determined by the underlying platform.

Each `ContainerizedWorkload` resource would need to maintain references to the traits that should apply to it in order for the controller that processes it to satisfy the above requirements of the contemporary OAM spec. Perhaps this would take the form of a `.spec.traitRefs` array, or perhaps we could model this using owner references (from trait to the component they apply to).

https://github.com/oam-dev/spec/blob/master/5.traits.md#traits-system-rules-and-characteristics

Reply:
Well, the deployment of traits only happens in AppConfig:

```yaml
kind: AppConfig
metadata: ...
spec:
  components:
  - componentName: ContainerizedWorkload
    instanceName: coolest-app-f92dm
    parameterValues: ...
    traits:
    - type: ManualScalerTrait
      properties:
        replicaCount: 3
```

Reply:
We're suggesting that an `ApplicationConfiguration` effectively be a template for instances of components and their traits - a little like how `KubernetesApplication` is a template for arbitrary Kubernetes resources (`Deployment`, etc). So under this pattern the OAM runtime (i.e. a Kubernetes controller) watching for `ApplicationConfiguration` would not actually deploy anything - instead it would create instances of each of the templated custom resource (in this case a `ContainerizedWorkload`, a `ManualScalerTrait`, and a `RedisCluster`. This would allow separate controllers to reconcile each of these custom resources. One advantage of this pattern is that it allows Crossplane's current resource claim controllers to work with OAM without a layer of abstraction (e.g. a `RedisCluster` workload type). What do you think?